File: backend/src/auth/jwt.ts
import jwt, { JwtPayload } from "jsonwebtoken";

import { SECRET_KEY, JWT_EXPIRATION } from "../config";
import { User } from "../schemas";
import { crudUser } from "../models";
import { ApiError, HttpStatus } from "../framework";

export interface UserTokenInput {
    userId: string;
    email: string;
}

export interface EncodedUserToken {
    userId: string;
    email: string;
    token: string;
}

export interface DecodedUserToken extends UserTokenInput, JwtPayload {}

export const createToken = (user: User): EncodedUserToken => {
    const payload: UserTokenInput = {
        userId: user.id,
        email: user.email,
    };
    const token = jwt.sign(payload, SECRET_KEY, { expiresIn: JWT_EXPIRATION });
    return { token, email: user.email, userId: user.id };
};

export const verifyToken = (token: string): DecodedUserToken => {
    const decoded = jwt.verify(token, SECRET_KEY);
    if (typeof decoded === "string") {
        throw new ApiError(HttpStatus.BAD_REQUEST, "Invalid token payload");
    }
    return decoded as DecodedUserToken;
};

export const getUserFromToken = async (token: string): Promise<User> => {
    const payload = verifyToken(token);
    const user = await crudUser.get(payload.userId);
    if (!user) {
        throw new ApiError(HttpStatus.BAD_REQUEST, "Invalid token payload");
    }
    if (user.email !== payload.email) {
        throw new ApiError(
            HttpStatus.BAD_REQUEST,
            "Invalid token, payload corrupted"
        );
    }
    return user;
};

File: backend-gin/internal/api/auth/jwt.go
package auth

import (
	"backend/internal/config"
	"backend/internal/lib/gin_"
	"backend/internal/lib/utils"
	"backend/internal/models/collections"
	"backend/internal/models/schemas"
	"context"
	"net/http"
	"strings"
)

func GetUserFromToken(token string) (schemas.UserRead, error) {
	badToken := gin_.ApiError{
		Code:    http.StatusUnauthorized,
		Message: "Token Not Valid",
	}

	payload, err := utils.DecodePayload(token, config.Env.SecretKey)
	if err != nil {
		if strings.Contains(err.Error(), "token expired") {
			return schemas.UserRead{}, gin_.ApiError{
				Code:    http.StatusUnauthorized,
				Message: "Token Expired",
			}
		}
		return schemas.UserRead{}, badToken
	}

	userIdRaw, userFound := payload["userId"]
	emailRaw, emailFound := payload["email"]
	if !userFound || !emailFound {
		return schemas.UserRead{}, badToken
	}

	userId, userIdValid := userIdRaw.(string)
	email, emailValid := emailRaw.(string)
	if !userIdValid || !emailValid {
		return schemas.UserRead{}, badToken
	}

	uc := collections.GetUserCollection()
	ctx := context.Background()
	user, err := uc.GetById(userId, ctx)
	if err != nil {
		return schemas.UserRead{}, gin_.ApiError{
			Code:    http.StatusNotFound,
			Message: "user not found",
		}
	}

	if user.Email != email {
		return schemas.UserRead{}, gin_.ApiError{
			Code:    http.StatusBadRequest,
			Message: "Invalid token, payload corrupted",
		}
	}

	return user, nil
}
